### PR TITLE
feat: Add ability to see messsage reactions (SQSERVICES-1876)

### PR DESCRIPTION
**Pull request by thisisamir98**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1876" title="SQSERVICES-1876" target="_blank">StorySQSERVICES-1876</a>  [Web] Link to and Detailed list of Reactions
  </td></table>
  <br />

### CODE REVIEW

File: src/script/components/input/ClassifiedBar.tsx [hunk 1]
```diff
@@ -37,31 +39,20 @@ interface ClassifiedBarProps {
...
     return null;
   }
   const classified = isClassified(users, classifiedDomains);
   const text = classified ? t('conversationClassified') : t('conversationNotClassified');
-  const highContrast = !classified;
+
   return (
-    <div data-uie-name="classified-label" css={{...barStyle(highContrast), ...style}}>
+    <div
+      className={cx('classified-bar', {green: classified, red: !classified})}
```
Comment (przemvs) [inline src/script/components/input/ClassifiedBar.tsx]: Please use for classes: **classified-bar--green** || **classified-bar--red** . Also for this usage you can use ternary operator **classified ? 'classified-bar--green' : 'classified-bar--red'**

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx [hunk 2]
```diff
@@ -47,15 +48,28 @@ import {PanelHeader} from '../PanelHeader';
...
 };
 
 const formatUserCount = (users: User[]): string => (users.length ? ` (${users.length})` : '');
 
+const getTotalReactionUsersCount = (reactions: Record<string, User[]>): number => {
+  let total = 0;
+  Object.keys(reactions).forEach(key => {
+    total += reactions[key].length;
+  });
+  return total;
```
Comment (przemvs) [inline src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx]: Use **Object.keys(reactions).reduce(fn, 0)**.

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx [hunk 3]
```diff
@@ -119,15 +133,32 @@ const MessageDetails: FC<MessageDetailsProps> = ({
...
+      const user = usersMap[userId];
+      const userReactions = reactions[userId] && reactions[userId].split(',');
+      for (const reaction of userReactions) {
+        if (reactionsGroupByUser[reaction]) {
+          reactionsGroupByUser[reaction].push(user);
+        } else {
+          reactionsGroupByUser[reaction] = [user];
+        }
+      }
+    }
```
Comment (przemvs) [inline src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx]: I'll use there Map object :)
This will be faster:

E.g ( Smth like that :) ): 

```
const usersMap = new Map();
usersReactions.forEach(user => {
    usersMap.set(user.id, user);
});

const reactionsGroupByUser = new Map();
for (const userId in reactions) {
      const user = usersMap.get(userId);
      const userReactions = reactions[userId]?.split(',');
      for (const reaction of userReactions) {
        if (reactionsGroupByUser.has(reaction)) {
          const currentUsers = reactionsGroupByUser.get(reaction);
          reactionsGroupByUser.set(reaction, [...currentUsers])
        } else {
          reactionsGroupByUser.set(reaction, [user]);
        }
      }
    }

```

File: server/package.json [hunk 4]
```diff
@@ -31,7 +31,7 @@
     "@types/jest": "^29.4.0",
     "@types/node": "18.11.18",
     "jest": "29.4.3",
-    "rimraf": "4.1.2",
+    "rimraf": "4.3.0",
```
Comment (atomrc) [inline server/package.json]: unrelated package updates should be done in a different PR. Please revert

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx (hunk 3 above)
Comment (atomrc) [inline src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx]: This seems to be duplicated code. I believe the `transformReactionObj` does exactly this, isn't it?

File: src/script/util/ReactionUtil.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+export function transformReactionObj(reactions: {[key: string]: string}) {
```
Comment (atomrc) [inline src/script/util/ReactionUtil.ts]: This naming is not clear. How about `groupByReaction`? WDYT?

**Review by atomrc — CHANGES_REQUESTED** [review]

File: src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx [hunk 6]
```diff
@@ -225,7 +212,7 @@ const ContentMessageComponent: React.FC<ContentMessageProps> = ({
       onMouseLeave={event => {
         // close floating message actions when no active menu is open like context menu/emoji picker
         if (!isMenuOpen) {
-          setActionMenuVisibility(false);
+          setActionMenuVisibility(true);
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx]: you're sure about this?

File: src/script/util/EmojiUtil.ts [hunk 7]
```diff
@@ -39,3 +42,35 @@ const isValidString = (string: string) => typeof string === 'string' && string.l
...
+});
+
+export const getEmojiTitleFromEmojiUnicode = (emojiUnicode: string): string => {
+  return emojiDictionary[emojiUnicode];
+};
+
+export function getEmojiUnicode(emojis: string) {
+  const hexCode = 16;
+  const padding = 4;
+  const unicode = [...emojis].map(emoji => emoji?.codePointAt(0)?.toString(hexCode).padStart(padding, '0')).join('-');
```
Comment (PatrykBuniX) [inline src/script/util/EmojiUtil.ts]: first `?` doesn't seem to be necessary, its always string

File: src/script/util/EmojiUtil.ts [hunk 8]
```diff
@@ -45,11 +45,11 @@ export const includesOnlyEmojis = (text: string) =>
 
 const emojiesFlatten = Object.values(emojies).flat();
 export const emojiesList = groupBy(emojiesFlatten, 'u');
-export const emojiDictionary: Record<string, string> = {};
+export const emojiDictionary: Map<string, string> = new Map();
 
 Object.keys(emojiesList).forEach(key => {
   // return if already existing in the dictionary
-  if (emojiDictionary[key]) {
+  if (emojiDictionary.get(key)) {
```
Comment (przemvs) [inline src/script/util/EmojiUtil.ts]:
```suggestion
  if (emojiDictionary.has(key)) {
```

**Review by PatrykBuniX — APPROVED** [review]

File: src/script/util/EmojiUtil.ts [hunk 9]
```diff
@@ -61,16 +61,16 @@ Object.keys(emojiesList).forEach(key => {
   const emojiObject = emojiValue[0];
   const emojiNames = emojiObject.n;
 
-  emojiDictionary[key] = emojiNames[emojiNames.length - 1].replaceAll('-', ' ');
+  emojiDictionary.set(key, emojiNames[0].replaceAll('-', ' '));
 });
 
 export const getEmojiTitleFromEmojiUnicode = (emojiUnicode: string): string => {
-  return emojiDictionary[emojiUnicode];
+  return emojiDictionary.get(emojiUnicode) ?? '';
```
Comment (przemvs) [inline src/script/util/EmojiUtil.ts]: Hmm.. I'm thinking on it: 
```suggestion
  return emojiDictionary.has(emojiUnicode) ? emojiDictionary.get(emojiUnicode) : '';
```

Need to compare performance of get and has :)

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx [hunk 10]
```diff
@@ -103,7 +118,7 @@ const MessageDetails: FC<MessageDetailsProps> = ({
   const teamId = activeConversation.team_id;
   const supportsReceipts = messageSender.isMe && teamId;
 
-  const supportsLikes = useMemo(() => {
+  const supportsReactions = useMemo(() => {
```
Comment (atomrc) [inline src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:123]: LEt's remove this `useMemo` it adds unecessary complexity here. memo should only be used for computation intensive operations

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx [hunk 11]
```diff
@@ -119,15 +134,28 @@ const MessageDetails: FC<MessageDetailsProps> = ({
...
+    reactionsGroupByUser.forEach((userIds, reaction) => {
+      reactionsGroupByUserMap.set(
+        reaction,
+        userIds.map(userId => usersMap.get(userId)!),
+      );
+    });
 
-    setLikeUsers(sortedUsersLikes);
+    setReactionUsers(reactionsGroupByUserMap);
   }, []);
```
Comment (atomrc) [inline src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:161]: LEt's remove the `useCallback` here, it unnecessary complexity and add more computation while rendering. 
Callbacks do not affect rendering performances

File: src/style/panel/message-details.less [hunk 12]
```diff
@@ -55,4 +55,19 @@
...
+
+    span:first-child {
+      margin-right: 5px;
+    }
+    span:last-child {
+      color: var(--gray-70);
+      font-size: @font-size-medium;
+      font-weight: @font-weight-bold;
+    }
+  }
```
Comment (atomrc) [inline src/style/panel/message-details.less]: Would it be possible to move this to a  `styles.ts` file next to the component?

**Review by atomrc — COMMENTED** [review]
A few things to consider, but looks good overall 💪

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx (hunk 11 above)
Comment (thisisamir98) [reply src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:161]: image

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx (hunk 10 above)
Comment (thisisamir98) [reply src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:123]: image

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx (hunk 11 above)
Comment (atomrc) [reply src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:161]: Bah ok, let's leave it for later then

File: src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx (hunk 10 above)
Comment (atomrc) [reply src/script/page/RightSidebar/MessageDetails/MessageDetails.tsx:123]: 🙃 nevermind, it's a problem for later then

File: src/style/panel/message-details.less (hunk 12 above)
Comment (thisisamir98) [reply src/style/panel/message-details.less]: https://github.com/wireapp/wire-webapp/pull/14825/commits/7f4d695a907dd251f66f43e71e0761144324341d

File: src/script/page/RightSidebar/MessageDetails/styles.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {CSSObject} from '@emotion/react';
+
+export const panelContentTitleStyles: CSSObject = {
+  paddingLeft: 20,
+  borderBottom: '1px solid var(--gray-50)',
```
Comment (V-Gira) [inline src/script/page/RightSidebar/MessageDetails/styles.ts]: didn't look at the specs but you could prob use `var(--border-color)` for this and `var(--text-input-placeholder)` for color to have css variable that respond to light and dark mode

Comment (thisisamir98) [reply src/script/page/RightSidebar/MessageDetails/styles.ts]: https://github.com/wireapp/wire-webapp/pull/14825/commits/4ef3e21cdbefcc7c98bc80d328d31e12d9b0413e

File: src/script/page/RightSidebar/MessageDetails/styles.ts
Comment (atomrc) [inline src/script/page/RightSidebar/MessageDetails/styles.ts:1]: Let's keep the same convention `MessageDetails.styles.ts`

**Review by V-Gira — APPROVED** [review]

**Review by atomrc — APPROVED** [review]

**Review by atomrc — APPROVED** [review]
